Tools/cameras_gimbals/siyi-download/siyi-download.py

```diff
@@ -0,0 +1,142 @@
+            dir_list = []
+            for dir in dir_dict_data_directories:
+                if ('path' in dir.keys()):
+                    dir_list.append(dir['path'])
+            print(prefix_str + f"{len(dir_list)} directories")
+
+            # get list of files in each directory
+            for dir_path in dir_list:
+                filenames_url = get_filelist_url(ip_address, media_type, dir_path)
+                with urlopen(filenames_url) as get_filenames_url:
+                    filename_dict = json.load(get_filenames_url)
+
+                    # check that the request succeeded
+                    if (not filename_dict['success']):
+                        exit(prefix_str + "failed to get list of files")
+
+                    # check response includes 'data'
+                    if ('data' not in filename_dict.keys()):
+                        exit(prefix_str + "could not get list of files, no 'data' in response")
+                    filename_dict_data = filename_dict['data']
+
+                    # check response includes 'list'
+                    if ('list' not in filename_dict_data.keys()):
+                        exit(prefix_str + "could not get list of files, no 'list' in response")
+                    filename_dict_data_list = filename_dict_data['list']
+                    print(prefix_str + f"{len(filename_dict_data_list)} files")
+
+                    # download each image
+                    for fileinfo in filename_dict_data_list:
+                        if ('name' not in fileinfo.keys() or 'url' not in fileinfo.keys()):
+                            exit(prefix_str + "could not get list of files, no 'name' or 'url' in response")
+                        filename = fileinfo['name']
+                        file_url = fileinfo['url']
+
+                        # correct incorrect ip address in returned url
+                        file_url_fixed = file_url.replace(ip_address_default, ip_address)
+
+                        # download file
+                        print(prefix_str + f"downloading {filename} from {file_url_fixed}")
+                        dest_filename = os.path.join(dest_dir, filename)
+                        try:
+                            urlretrieve(file_url_fixed, dest_filename)
+                        except Exception as e:
+                            print(prefix_str + f"failed to download {filename}: {e}")
+
+
+# main function
+def main():
+    parser = ArgumentParser(description=__doc__)
+    parser.add_argument("--ipaddr", default=ip_address_default, help="IP address of camera")
+    parser.add_argument("--dest", default=".", help="destination directory where downloaded files will be saved")
+    args = parser.parse_args()
+
+    # check destination directory exists
+    if not os.path.exists(args.dest):
+        exit(prefix_str + "invalid destination directory")
+
+    # download files
+    download_files(args.ipaddr, args.dest)
```

Comment:
```suggestion
    downloader = Downloader(args.ipaddr, args.dest)
    downloader.run()
```

Doesn't need to be done (too big a change), but I've found that this structure is where you want to end up anyway.

Reply:
OK noted, for now I'll leave it as a more functional scripts but in the future I'll tend towards making classes